compiler/src/dotty/tools/dotc/core/PatternTypeConstrainer.scala

```diff
@@ -264,28 +264,19 @@ trait PatternTypeConstrainer { self: TypeComparer =>
     trace(i"constraining simple pattern type $tp >:< $pt", gadts, (res: Boolean) => i"$res gadt = ${ctx.gadt}") {
       (tp, pt) match {
         case (AppliedType(tyconS, argsS), AppliedType(tyconP, argsP)) =>
-          val saved = state.nn.constraint
-          val result =
-            ctx.gadt.rollbackGadtUnless {
-              tyconS.typeParams.lazyZip(argsS).lazyZip(argsP).forall { (param, argS, argP) =>
-                val variance = param.paramVarianceSign
-                if variance == 0 || assumeInvariantRefinement ||
-                  // As a special case, when pattern and scrutinee types have the same type constructor,
-                  // we infer better bounds for pattern-bound abstract types.
-                  argP.typeSymbol.isPatternBound && patternTp.classSymbol == scrutineeTp.classSymbol
-                then
-                  val TypeBounds(loS, hiS) = argS.bounds
-                  val TypeBounds(loP, hiP) = argP.bounds
-                  var res = true
-                  if variance <  1 then res &&= isSubType(loS, hiP)
-                  if variance > -1 then res &&= isSubType(loP, hiS)
-                  res
-                else true
-              }
-            }
-          if !result then
-            constraint = saved
-          result
+          tyconS.typeParams.lazyZip(argsS).lazyZip(argsP).foreach { (param, argS, argP) =>
+            val variance = param.paramVarianceSign
+            if variance == 0 || assumeInvariantRefinement ||
+              // As a special case, when pattern and scrutinee types have the same type constructor,
+              // we infer better bounds for pattern-bound abstract types.
+              argP.typeSymbol.isPatternBound && patternTp.classSymbol == scrutineeTp.classSymbol
+            then
+              val TypeBounds(loS, hiS) = argS.bounds
+              val TypeBounds(loP, hiP) = argP.bounds
+              if variance <  1 then isSubType(loS, hiP)
+              if variance > -1 then isSubType(loP, hiS)
+          }
+          true
```

Comment:
I think we should still return the results from `TypeComparer` here, since it is used [here](https://github.com/lampepfl/dotty/blob/main/compiler/src/dotty/tools/dotc/core/PatternTypeConstrainer.scala#L180) here by `either`.